### PR TITLE
expm implementation for dense matrices

### PR DESCRIPTION
**Pull request by MrRobot2211**
This will implement the equivalent to `scipy.linalg.expm` for dense matrices.   

Implemetationss worth taking into account, both [TensorFlow](https://www.tensorflow.org/api_docs/python/tf/linalg/expm) and [PyTorch](https://pytorch.org/docs/stable/generated/torch.matrix_exp.html) already have implementations. The former's relies on scaling and squaring Padé's aproximation while the latter uses an optimized polynomial taylor sum approximation.

There is also [this](https://github.com/pd93/CUDAMatrixExp) implementation, which is also based on Pade's aproximation.

And this reference from the implementation on [Mathematica](https://www.mathworks.com/matlabcentral/fileexchange/67668-batch-matrix-exponential-computation-with-cpu-gpu), which hints at some optimizations for hermitian matrices.

This will tackle #282, and help with https://github.com/qutip/qutip-cupy/issues/40 .

### CODE REVIEW

**Comment by kmaehashi** [comment]
Thank you for working on this! I think `cupyx.scipy.linalg.expm` is the right place to put this function as `numpy.linalg.expm` does not exist 😄

https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.expm.html

SciPy's implementation supports both dense and sparse but it's ok to start with the dense-only implementation.

**Comment by MrRobot2211** [comment]
> Thank you for working on this! I think `cupyx.scipy.linalg.expm` is the right place to put this function as `numpy.linalg.expm` does not exist 
> 
> https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.expm.html
> 
> SciPy's implementation supports both dense and sparse but it's ok to start with the dense-only implementation.

Thank you for the catch , I am editing the description and changing the locationof the files.

File: .gitignore [hunk 1]
```diff
@@ -18,6 +18,7 @@ cupy.egg-info/
 dist/
 htmlcov/
 .idea/
+.vscode/
```
Comment (takagi) [inline .gitignore:21]: We do not have `.gitignore` setting for development environments. Would you remove this change from the PR? Instead, you can use `~/.config/git/ignore` to ignore vscode files in all situations.

**Comment by leofang** [comment]
Hi @MrRobot2211 Any chance you can push your WIP so that I can ask someone to take over and help move it pass the finish line? As you know this is an important API for quantum computing and quantum physics, so let's not drop the ball 🙂 Thanks!

**Comment by kmaehashi** [comment]
Kindly ping @MrRobot2211 😃

**Comment by leofang** [comment]
So I think this repo (https://github.com/Lezcano/expm) was how PyTorch used to implement `expm`. It implements both GPU-friendly approaches (see its README and the linked papers). We could port it to CuPy and see how the performance goes, and decide if we need any dedicated CUDA kernel. Could be a good mini project 🙂

**Comment by ev-br** [comment]
https://github.com/cupy/cupy/pull/7660 has an implementation based on Pade 13/13 approximation of Higham'2005

**Comment by leofang** [comment]
Thanks, @ev-br, it's very encouraging to know! I suggest we close this PR due to inactivity. Once #7660 is in, we can revisit `expm()` anytime if the Taylor-expansion based approach is needed.

**Comment by ev-br** [comment]
If/when this PR is closed, I'd suggest to open an enhancement issue with links to other implementations, especially Lezcano's one (https://github.com/cupy/cupy/pull/5688#issuecomment-1552067121). As a general purpose expm for potentially large matrices, it is probably way more efficient than gh-7660's version. And as a yet further enhancement, https://github.com/higham/expmv has a Taylor-expansion based expm times vector algorithm.

**Comment by leofang** [comment]
@takagi shall we close this PR and follow up in #7769?

**Comment by takagi** [comment]
Yes!

**Pull request closed without merging**